File: onyx-database-tests/test/remote/callbacks/EntityWithCallbacksTest.java
package remote.callbacks;

import com.onyx.exception.OnyxException;
import com.onyx.exception.InitializationException;
import entities.EntityWithCallbacks;
import entities.SequencedEntityWithCallbacks;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import remote.base.RemoteBaseTest;

import java.io.IOException;

/**
 * Created by Chris Osborn on 12/29/2014.
 */
public class EntityWithCallbacksTest extends RemoteBaseTest {


    @Before
    public void before() throws InitializationException, InterruptedException
    {
        initialize();
    }

    @After
    public void after() throws IOException
    {
        shutdown();
    }

    @Test
    public void testPrePersistCallbacksForHashIndex() throws OnyxException {
        //Create new entity
        EntityWithCallbacks entity = new EntityWithCallbacks();
        entity.setId("1");
        entity.setName("INSERT");
        manager.saveEntity(entity);

        //retrieve the entity
        EntityWithCallbacks savedEntity = new EntityWithCallbacks();
        savedEntity.setId("1");
        manager.find(savedEntity);

        //Update the entity
        savedEntity.setName(savedEntity.getName() + "&UPDATE");
        manager.saveEntity(savedEntity);

        //Assert
        Assert.assertTrue(savedEntity.getName().contains("INSERT"));
        Assert.assertTrue(savedEntity.getName().contains("UPDATE"));
        Assert.assertTrue(savedEntity.getName().contains("_PrePersist"));
        Assert.assertTrue(savedEntity.getName().contains("_PreInsert"));
        Assert.assertTrue(savedEntity.getName().contains("_PreUpdate"));
    }

    @Test
    public void testPrePersistCallbacksForSequenceIndex() throws OnyxException {
        //Create new entity
        SequencedEntityWithCallbacks entity = new SequencedEntityWithCallbacks();
        entity.setName("INSERT");
        entity = (SequencedEntityWithCallbacks) manager.saveEntity(entity);

        //retrieve the entity
        SequencedEntityWithCallbacks savedEntity = (SequencedEntityWithCallbacks) manager.find(entity);

        //Update the entity
        savedEntity.setName(savedEntity.getName() + "&UPDATE");
        manager.saveEntity(savedEntity);

        //Assert
        Assert.assertTrue(savedEntity.getName().contains("INSERT"));
        Assert.assertTrue(savedEntity.getName().contains("UPDATE"));
        Assert.assertTrue(savedEntity.getName().contains("_PrePersist"));
        Assert.assertTrue(savedEntity.getName().contains("_PreInsert"));
        Assert.assertTrue(savedEntity.getName().contains("_PreUpdate"));
    }

    @Test
    public void testPostPersistCallbacksForHashIndex() throws OnyxException {
        //Create new entity
        EntityWithCallbacks entity = new EntityWithCallbacks();
        entity.setId("1");
        entity.setName("INSERT");

        try
        {
            manager.deleteEntity(entity);

        } catch (OnyxException e)
        { }

        manager.saveEntity(entity);

        Assert.assertTrue(entity.getName().contains("_PostInsert"));

        //retrieve the entity
        EntityWithCallbacks savedEntity = new EntityWithCallbacks();
        savedEntity.setId("1");
        manager.find(savedEntity);

        //Update the entity
        savedEntity.setName(savedEntity.getName() + "&UPDATE");
        manager.saveEntity(savedEntity);

        //Assert
        Assert.assertTrue(savedEntity.getName().contains("INSERT"));
        Assert.assertTrue(savedEntity.getName().contains("UPDATE"));
        Assert.assertTrue(savedEntity.getName().contains("_PostPersist"));
        Assert.assertTrue(savedEntity.getName().contains("_PostUpdate"));
    }

    @Test
    public void testPostPersistCallbacksForSequenceIndex() throws OnyxException {
        //Create new entity
        SequencedEntityWithCallbacks entity = new SequencedEntityWithCallbacks();
        entity.setName("INSERT");
        entity = (SequencedEntityWithCallbacks) manager.saveEntity(entity);

        Assert.assertTrue(entity.getName().contains("_PostInsert"));

        //retrieve the entity
        SequencedEntityWithCallbacks savedEntity = (SequencedEntityWithCallbacks) manager.find(entity);

        //Update the entity
        savedEntity.setName(savedEntity.getName() + "&UPDATE");
        manager.saveEntity(savedEntity);

        //Assert
        Assert.assertTrue(savedEntity.getName().contains("INSERT"));
        Assert.assertTrue(savedEntity.getName().contains("UPDATE"));
        Assert.assertTrue(savedEntity.getName().contains("_PostPersist"));
        Assert.assertTrue(savedEntity.getName().contains("_PostUpdate"));
    }

}

File: onyx-database-tests/test/remote/callbacks/EntityWithCallbacksTest.kt
package remote.callbacks

import com.onyx.exception.OnyxException
import com.onyx.exception.InitializationException
import com.onyx.persistence.IManagedEntity
import entities.EntityWithCallbacks
import entities.SequencedEntityWithCallbacks
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import remote.base.RemoteBaseTest

import java.io.IOException

/**
 * Created by Chris Osborn on 12/29/2014.
 */
class EntityWithCallbacksTest : RemoteBaseTest() {


    @Before
    @Throws(InitializationException::class, InterruptedException::class)
    fun before() {
        initialize()
    }

    @After
    @Throws(IOException::class)
    fun after() {
        shutdown()
    }

    @Test
    @Throws(OnyxException::class)
    fun testPrePersistCallbacksForHashIndex() {
        //Create new entity
        val entity = EntityWithCallbacks()
        entity.id = "1"
        entity.name = "INSERT"
        manager!!.saveEntity<IManagedEntity>(entity)

        //retrieve the entity
        val savedEntity = EntityWithCallbacks()
        savedEntity.id = "1"
        manager!!.find<IManagedEntity>(savedEntity)

        //Update the entity
        savedEntity.name = savedEntity.name!! + "&UPDATE"
        manager!!.saveEntity<IManagedEntity>(savedEntity)

        //Assert
        Assert.assertTrue(savedEntity.name!!.contains("INSERT"))
        Assert.assertTrue(savedEntity.name!!.contains("UPDATE"))
        Assert.assertTrue(savedEntity.name!!.contains("_PrePersist"))
        Assert.assertTrue(savedEntity.name!!.contains("_PreInsert"))
        Assert.assertTrue(savedEntity.name!!.contains("_PreUpdate"))
    }

    @Test
    @Throws(OnyxException::class)
    fun testPrePersistCallbacksForSequenceIndex() {
        //Create new entity
        var entity = SequencedEntityWithCallbacks()
        entity.name = "INSERT"
        entity = manager!!.saveEntity(entity)

        //retrieve the entity
        val savedEntity = manager!!.find<IManagedEntity>(entity) as SequencedEntityWithCallbacks

        //Update the entity
        savedEntity.name = savedEntity.name + "&UPDATE"
        manager!!.saveEntity<IManagedEntity>(savedEntity)

        //Assert
        Assert.assertTrue(savedEntity.name.contains("INSERT"))
        Assert.assertTrue(savedEntity.name.contains("UPDATE"))
        Assert.assertTrue(savedEntity.name.contains("_PrePersist"))
        Assert.assertTrue(savedEntity.name.contains("_PreInsert"))
        Assert.assertTrue(savedEntity.name.contains("_PreUpdate"))
    }

    @Test
    @Throws(OnyxException::class)
    fun testPostPersistCallbacksForHashIndex() {
        //Create new entity
        val entity = EntityWithCallbacks()
        entity.id = "1"
        entity.name = "INSERT"

        try {
            manager!!.deleteEntity(entity)

        } catch (e: OnyxException) {
        }

        manager!!.saveEntity<IManagedEntity>(entity)

        Assert.assertTrue(entity.name!!.contains("_PostInsert"))

        //retrieve the entity
        val savedEntity = EntityWithCallbacks()
        savedEntity.id = "1"
        manager!!.find<IManagedEntity>(savedEntity)

        //Update the entity
        savedEntity.name = savedEntity.name!! + "&UPDATE"
        manager!!.saveEntity<IManagedEntity>(savedEntity)

        //Assert
        Assert.assertTrue(savedEntity.name!!.contains("INSERT"))
        Assert.assertTrue(savedEntity.name!!.contains("UPDATE"))
        Assert.assertTrue(savedEntity.name!!.contains("_PostPersist"))
        Assert.assertTrue(savedEntity.name!!.contains("_PostUpdate"))
    }

    @Test
    @Throws(OnyxException::class)
    fun testPostPersistCallbacksForSequenceIndex() {
        //Create new entity
        var entity = SequencedEntityWithCallbacks()
        entity.name = "INSERT"
        entity = manager!!.saveEntity(entity)

        Assert.assertTrue(entity.name.contains("_PostInsert"))

        //retrieve the entity
        val savedEntity = manager!!.find<IManagedEntity>(entity) as SequencedEntityWithCallbacks

        //Update the entity
        savedEntity.name = savedEntity.name + "&UPDATE"
        manager!!.saveEntity<IManagedEntity>(savedEntity)

        //Assert
        Assert.assertTrue(savedEntity.name.contains("INSERT"))
        Assert.assertTrue(savedEntity.name.contains("UPDATE"))
        Assert.assertTrue(savedEntity.name.contains("_PostPersist"))
        Assert.assertTrue(savedEntity.name.contains("_PostUpdate"))
    }

}
